crates/server/src/rpc_handler.rs

```diff
@@ -0,0 +1,128 @@
+use std::fmt::{self};
+
+use axum::{
+    extract::{rejection::JsonRejection, Extension},
+    Json,
+};
+use futures::{future, FutureExt};
+use rpc_core::{
+    error::RpcError,
+    request::{Request, RpcCall, RpcMethodCall},
+    response::{Response, ResponseResult, RpcResponse},
+};
+use serde::de::DeserializeOwned;
+use tracing::{error, trace, warn};
+
+/// Helper trait that is used to execute starknet rpc calls
+#[async_trait::async_trait]
+pub trait RpcHandler: Clone + Send + Sync + 'static {
+    /// The request type to expect
+    type Request: DeserializeOwned + Send + Sync + fmt::Debug;
+
+    /// Invoked when the request was received
+    async fn on_request(&self, request: Self::Request) -> ResponseResult;
+
+    /// Invoked for every incoming `RpcMethodCall`
+    ///
+    /// This will attempt to deserialize a `{ "method" : "<name>", "params": "<params>" }` message
+    /// into the `Request` type of this handler. If a `Request` instance was deserialized
+    /// successfully, [`Self::on_request`] will be invoked.
+    ///
+    /// **Note**: override this function if the expected `Request` deviates from `{ "method" :
+    /// "<name>", "params": "<params>" }`
+    async fn on_call(&self, call: RpcMethodCall) -> RpcResponse {
+        trace!(target: "rpc",  id = ?call.id , method = ?call.method, "received method call");
+        let RpcMethodCall {
+            method, params, id, ..
+        } = call;
+
+        let params: serde_json::Value = params.into();
+        let call = serde_json::json!({
+            "method": &method,
+            "params": params
+        });
+
+        match serde_json::from_value::<Self::Request>(call) {
+            Ok(req) => {
+                let result = self.on_request(req).await;
+                RpcResponse::new(id, result)
+            }
+            Err(err) => {
+                let err = err.to_string();
+                if err.contains("unknown variant") {
```

Comment:
I thought hackiness like this was not possible in Rust. Can't we distinguish between cases based on something else, e.g. error type?

Reply:
Its easier to check the string of the error

Reply:
When deserialising enums that are marked with #[serde(untagged)] https://serde.rs/enum-representations.html. If it is not able to deserialize to any variant of the Enum it returns the error unknown variant.